File: pset6/credit/credit.py
from cs50 import get_int
from math import floor


def main():
    card_number = get_int("Number: ")

    if (is_valid(card_number)):
        print(get_provider(card_number))
    else:
        print("INVALID\n")


def is_valid(card_number):
    digits = double_alternate_digits(card_number)
    sum = sum_elements_by_digit(digits)
    return (sum % 10 == 0)


def double_alternate_digits(number):
    num_as_string = str(number)

    result = []
    position = len(num_as_string) - 1
    counter = 0
    while (position >= 0):
        digit = int(num_as_string[position])

        if (counter % 2):
            digit *= 2

        result.append(digit)

        position -= 1
        counter += 1

    return result


def sum_elements_by_digit(input):
    sum = 0
    for element in input:
        if element < 10:
            sum += element
        else:
            string = str(element)
            for digit in string:
                sum += int(digit)

    return sum


def get_provider(raw_card_number):
    starts_with = int(str(raw_card_number)[:2])
    length = len(str(raw_card_number))

    if (starts_with == 34 or starts_with == 37) and length == 15:
        return "AMEX"
    elif (40 <= starts_with and starts_with <= 49 and
          (length == 13 or length == 16)):
        return "VISA"
    elif (51 <= starts_with and starts_with <= 55) and (length == 16):
        return "MASTERCARD"
    else:
        return "INVALID"


if __name__ == '__main__':
    main()

File: pset1/credit/more/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int countDigits(long long x)
{
    return floor(log10(x)) + 1;
}

void getArrayFromNumber(int arr[], int size, long long number)
{
    for (int i = 0; i < size; i++)
    {
        int digit = number % 10;
        arr[size - i - 1] = digit;
        number /= 10;
    }
}

void doubleAlternateDigits(int card_number[], int size)
{
    int index_to_double = size - 2;
    int j = 0;

    while (index_to_double > -1)
    {
        card_number[index_to_double] *= 2;
        index_to_double -= 2;
        j += 1;
    }
}

int arraySum(int arr[], int size)
{
    int sum = 0;
    for (int i = 0; i < size; i++)
    {
        int element = arr[i];
        if (element > 9)
        {
            int sub_elements[2];
            getArrayFromNumber(sub_elements, 2, element);
            sum += sub_elements[0];
            sum += sub_elements[1];
        }
        else
        {
            sum += arr[i];
        }
    }
    return sum;
}

bool isValid(int sum)
{
    return sum % 10 == 0;
}

string getProvider(long long raw_card_number, int size)
{
    long long first_digits = raw_card_number;
    while (first_digits > 99)
    {
        first_digits /= 10;
    }

    if ((first_digits == 34 || first_digits == 37) && size == 15)
    {
        return "AMEX\n";
    }
    else if ((40 <= first_digits && first_digits <= 49) && (size == 13 || size == 16))
    {
        return "VISA\n";
    }
    else if ((51 <= first_digits && first_digits <= 55) && (size == 16))
    {
        return "MASTERCARD\n";
    }
    else
    {
        return "INVALID\n";
    }
}



int main(void)
{
    long long raw_card_number = get_long_long("Number: ");
    int size = countDigits(raw_card_number);
    string provider = getProvider(raw_card_number, size);

    int card_number[size];
    getArrayFromNumber(card_number, size, raw_card_number);

    doubleAlternateDigits(card_number, size);

    int sum = arraySum(card_number, size);

    if (isValid(sum))
    {
        printf("%s", provider);
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}